standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MySQLConnectorProvider.java

```diff
@@ -90,10 +90,20 @@ protected String getDataType(String dbDataType, int size) {
     // map any db specific types here.
     switch (dbDataType.toLowerCase())
     {
+    case "bit":
+      return toHiveBitType(size);
     default:
       mappedType = ColumnType.VOID_TYPE_NAME;
       break;
     }
     return mappedType;
   }
+
+  private String toHiveBitType(int size) {
+    if (size <= 1) {
+      return ColumnType.BOOLEAN_TYPE_NAME;
+    } else {
+      return ColumnType.BIGINT_TYPE_NAME;
```

Comment:
so bit datatype in MYSQL stores and array of bits. for any column, bit-type, with size > 1, the proposal is to use bigint. If I understand this correctly, b'111000' will be read as '111000' (numerical). Isnt this incorrect as these are not equivalent? But also to be able to read this datatype, should we read as "select bin(col)" while reading from remote table? Can you give me an example of how you tested this and what the results were? Thank you

Reply:
for bit-type with size <= 1 of mysql column, users usually use this column to express false or true, so it make sense to be recognized boolean type in hive.

for bit-type with size >1 of mysql column, users may want to express different semantics, eg weekly working calendar, bit(7) , and in this case we can not simply convert bit(7) to boolean type. I convert this bit(n) to bigint in hive, eg **b'111000' will be read as 56**.

in this pr, i  added qtest to select bit datatype, but HIVE-26192 blocked me, i hope you can give advice about HIVE-26192。

>  But also to be able to read this datatype, should we read as "select bin(col)" while reading from remote table?
I think it is diffcult and not worth to read as "select bin(col)".  Becaue if we want do this, we may need to mak type check when executing query and then convert column type to bit.  Also, hive can not recognize bit type, and i think hive will compile failed with bit type before the query is submited to mysql.  That's just my basic thought. 
 
In addtion, i have checked  the presto and spark code. presto will convert all bit type to bolean. Spark have the similar ideas with this pr. I think spark idea is more reasonable.
https://github.com/trinodb/trino/blob/543ae143cadeb47ab03af4197dae9d00ff5baf7c/plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java#L377-L379

https://github.com/apache/spark/blob/7309e76d8b95e306d6f3d2f611316b748949e9cf/sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala#L64-#L71

Reply:
> for BIT(n), if we represent the value as bigint(decimal) in hive, will the original intent have been lost? a select would return 56 which visually doesn't represent how it is originally stored

I think users migiht be able to convert the result **56** to bit **"111000"** themselves instead hive internally.  :)

I will research how to do convert result using hive functions or UDFs as you suggsted. or maybe we can rewrite query sql to "select bin(col)" after hive compile.

Reply:
@zhangbutao Any luck with the above item related to bit datatype

Reply:
@nrg4878 Sorry for not updating the issue for a long time. 
In my opinion, we have two ways to get origina MySQL BITs values, that is b'111000' in MySQL will be read as '111000'(binary number) in hive.

First way:
In MySQL, we can get bit values using 'select bin(days) from work_calendar'. So i was thinking how we use MySQL native language to get bit values in hive. After some research, I found we could use jdbc table computation pushdown(hive.sql.query) to rewrite query sql. That is to say, when we query a bit type table using 'select days from work_calendar' in hive, we can rewrite the pushdown query to 'select bin(days) from work_calendar'. In this way, we may need check original MySQL data type and rewrite bit type query when pushdown. Maybe we should do some change in JdbcSerde.java.
https://cwiki.apache.org/confluence/display/Hive/JDBC+Storage+Handler#JDBCStorageHandler-ComputationPushdown

Second way:
I was also thinking using udf to to cast valus to BITs as you suggested. That is, 56(bigint) is cast to '111000'(binary number) in hive internal. Fortunately, I found hive already has this udf bin:
https://github.com/apache/hive/blob/f29cb2245c97102975ea0dd73783049eaa0947a0/ql/src/java/org/apache/hadoop/hive/ql/udf/UDFBin.java#L28-L37
So, based on this pr, we can get bit values using hive query 'select bin(days) from work_calendar'. And we no need change current jdbc code(JdbcSerDe.java or others). I like this simple way as we can reuse bin udf to acheive our goal.

wdyt? if you have better ways, feel free to come up with. thank you.

Reply:
@zhangbutao Sorry for a delayed response. 
So with option #1, HMS would replace the bit type columns for MySQL with bin(column_name). I kinda like this option in that it simplifies the logic and pushes the onus on the JDBC datasource natively. 
Option #2 uses hive UDF which maybe different from the native implementation. And it also involves computation on the hive side.
What do you think?

Reply:
@nrg4878 I agree option 1 is the best and native way to handle datasouce specific datatype. But there is need to explore more info about JdbcStorageHandler, e.g. Jdbc pushdown or sth else.
Anyway, I will try to research and implement option 1.  I will update here if there is any progress. Thx.

Reply:
Can we convert the bit(n>1) to hive binary type?

Reply:
@zhangbutao is option #1 above not possible where we set the hive.sql.query to include like a select bin(column_name) ... on the table containing bit type columns?

mapping to hive binary type might work. One concern is that doesn't hive expect binary values to be within ''. So something like '0111' instead of 0111. Can you please try it out to see how this works end to end?

Reply:
@dengzhhu653 @nrg4878 Thanks for your suggestion.
Actually, i have tried to convert MySQL's bit datatype to hive's binary, but did not get expected restult. In this case, hive stores MySQL's  bit values(b'111000') as binary type, and this binary values can not dispaly to users, that is to say users only can see a blank result.

Step to test:
1. create mysql table with bit data type in mysql:
![image](https://user-images.githubusercontent.com/9760681/187333226-f23705a5-50b6-40bf-b1e1-8fcecc7a065c.png)

2. create jdbc-mapping table in hive:
`   CREATE EXTERNAL TABLE jdbc_testmysqlbit`
`(`
`  id binary`
`)`
`STORED BY 'org.apache.hive.storage.jdbc.JdbcStorageHandler'`
`TBLPROPERTIES (`
`"hive.sql.database.type" = "MYSQL",`
`"hive.sql.jdbc.driver" = "com.mysql.jdbc.Driver",`
`"hive.sql.jdbc.url" = "jdbc:mysql://localhost:3306/testdb",`
`"hive.sql.dbcp.username" = "user",`
`"hive.sql.dbcp.password" = "passwd",`
`"hive.sql.table" = "testmysqlbit",`
`"hive.sql.dbcp.maxActive" = "1"`
`);`

3.`select * from jdbc_testmysqlbit;`  using hive beeline:
![image](https://user-images.githubusercontent.com/9760681/187333975-e7d6270c-e922-4088-948f-46024cbcb6e5.png)

Reply:
> option 1 above not possible where we set the hive.sql.query to include like a select bin(column_name) ... on the table containing bit type columns?

@nrg4878 I have tried set the hive.sql.query to include query like` select bin(id) from testmysqlbit,` and it worked as expected. That is to say, hive.sql.query can push down MySQL's native query to remote MySQL datasouce, and we can get MySQL's bit datatype values using a visible hive's bitint type or hive's string type.

Step to test:
1. create jdbc-mapping table with **hive.sql.query** in hive:
`   CREATE EXTERNAL TABLE jdbc_testmysqlbit_with_query`
`(`
`  id bigint`
`)`
`STORED BY 'org.apache.hive.storage.jdbc.JdbcStorageHandler'`
`TBLPROPERTIES (`
`"hive.sql.database.type" = "MYSQL",`
`"hive.sql.jdbc.driver" = "com.mysql.jdbc.Driver",`
`"hive.sql.jdbc.url" = "jdbc:mysql://localhost:3306/testdb",`
`"hive.sql.dbcp.username" = "root",`
`"hive.sql.dbcp.password" = "passwd",`
`"hive.sql.table" = "testmysqlbit",`
`"hive.sql.dbcp.maxActive" = "1",`
`"hive.sql.query" = "select bin(id) from testmysqlbit"`
`);`

2. `select * from jdbc_testmysqlbit_with_query;` using hive beeline:
![image](https://user-images.githubusercontent.com/9760681/187336443-51942986-1284-4fc2-90b8-eae02db3beb2.png)

Reply:
@nrg4878 Another alternative is the one I mentioned earlier, that is we use hive's interal udf bin() to convert the bigint values to bit type. This one needs users to be aware of hive's bin() udf:

Step to test:
1. Create jdbc-mapping table in hive. Use hive's bigint to map MySQL's bit datatype:
 ` CREATE EXTERNAL TABLE jdbc_testmysqlbit_use_bigint`
`(`
`  id bigint`
`)`
`STORED BY 'org.apache.hive.storage.jdbc.JdbcStorageHandler'`
`TBLPROPERTIES (`
`"hive.sql.database.type" = "MYSQL",`
`"hive.sql.jdbc.driver" = "com.mysql.jdbc.Driver",`
`"hive.sql.jdbc.url" = "jdbc:mysql://localhost:3306/testdb",`
`"hive.sql.dbcp.username" = "user",`
`"hive.sql.dbcp.password" = "passwd",`
`"hive.sql.table" = "testmysqlbit",`
`"hive.sql.dbcp.maxActive" = "1"`
`);`
 
2. `select * from jdbc_testmysqlbit_use_bigint;` using hive beeline:
![image](https://user-images.githubusercontent.com/9760681/187339777-16cc48f7-af07-4e31-8095-660f8e8e0c58.png)

3. `select bin(id) from jdbc_testmysqlbit_use_bigint;` using hive beeline:
  
![image](https://user-images.githubusercontent.com/9760681/187339967-bd33c930-85de-4da5-ae4c-8a3f13d7763d.png)

Reply:
Converting to bigint may have problems in joining or pushing predicate down on the bit column. I wondering if we take the trino way to deal with such column: converts the bit(1) to boolean type, and for those greater than 1, throws MetaException indicating that Hive is unable to map that remote table.

Reply:
Not sure if this way(to bigint) affects the joining or pushing predicate down in hive. I have checked Spark related code earlier and that's what spark did.

The Trino's way is simple and straightforward, but maybe lose the real meaning of MySQL's bit(n>1). However, i also think this is an optional solution because maybe we don't want  spare much time to deal with the special and not commonly used data type.